File: eval-reviews/2026-09-12-chained/query-null/typescript/fable/r1/starter/main.ts
import { readFileSync } from "node:fs";
import { DomainError, validate, requireThat } from "./model.ts";
import { Parser } from "./parser.ts";
import { bind } from "./binder.ts";
import { optimize, execute } from "./engine.ts";
let response: unknown;
try {
  const [database, queries] = validate(
    JSON.parse(
      readFileSync(0, "utf8"),
      (_key: string, value: unknown, context?: { source?: string }) => {
        requireThat(
          typeof value !== "number" || !/[.eE]/.test(context?.source ?? ""),
          "INVALID_INPUT",
        );
        return value;
      },
    ) as unknown,
  );
  const results = queries.map((q) => {
    const plan = bind(new Parser(q.sql).parse(), database);
    return execute(q.optimize ? optimize(plan) : plan);
  });
  response = { ok: true, result: { results } };
} catch (e) {
  if (!(e instanceof DomainError) && !(e instanceof SyntaxError)) throw e;
  response = {
    ok: false,
    error: { code: e instanceof DomainError ? e.message : "INVALID_INPUT" },
  };
}
process.stdout.write(JSON.stringify(response) + "\n");

File: eval-reviews/2026-09-12-chained/query-null/python/fable/r1/starter/main.py
#!/usr/bin/env python3
import json
import sys
from model import DomainError, validate
from parser import Parser
from binder import bind
from engine import optimize, execute
from views import Store


def main():
    try:
        request = json.loads(sys.stdin.read())
        database, queries, commands = validate(request)
        if commands is not None:
            store = Store(database)
            results = [store.run(command) for command in commands]
        else:
            results = []
            for query in queries:
                plan = bind(Parser(query['sql']).parse(), database)
                results.append(execute(optimize(plan) if query['optimize'] else plan))
        response = {'ok': True, 'result': {'results': results}}
    except DomainError as e:
        response = {'ok': False, 'error': {'code': str(e)}}
    except (ValueError, KeyError, TypeError):
        response = {'ok': False, 'error': {'code': 'INVALID_INPUT'}}
    print(json.dumps(response, separators=(',', ':')))

if __name__ == '__main__': main()
